src/transformers/trainer.py

```diff
@@ -93,6 +94,44 @@ def torch_distributed_zero_first(local_rank: int):
         torch.distributed.barrier()
 
 
+class SequentialDistributedSampler(Sampler):
+    """
+    Distributed Sampler that subsamples indicies sequentially,
+    making it easier to collate all results at the end.
+    """
+
+    def __init__(self, dataset, num_replicas=None, rank=None):
+        if num_replicas is None:
+            if not torch.distributed.is_available():
+                raise RuntimeError("Requires distributed package to be available")
+            num_replicas = torch.distributed.get_world_size()
+        if rank is None:
+            if not torch.distributed.is_available():
+                raise RuntimeError("Requires distributed package to be available")
+            rank = torch.distributed.get_rank()
+        self.dataset = dataset
+        self.num_replicas = num_replicas
+        self.rank = rank
+        self.num_samples = int(math.ceil(len(self.dataset) * 1.0 / self.num_replicas))
+        self.total_size = self.num_samples * self.num_replicas
+
+    def __iter__(self):
+        indices = list(range(len(self.dataset)))
+
+        # add extra samples to make it evenly divisible
+        indices += indices[: (self.total_size - len(indices))]
+        assert len(indices) == self.total_size
+
+        # subsample
+        indices = indices[self.rank * self.num_samples : (self.rank + 1) * self.num_samples]
```

Comment:
This is the main difference with `DistributedSampler`

Reply:
Can you explain the reasoning behind making this sequential, i.e. unshuffled? I know that the way you've implemented the gathering part forces you to use a sequential distributed sampler, but do you foresee an event where you may need to shuffle the inputs? 

IMO, the ideal solution would be to implement this exactly the same as distributed sampler except for the part where we add samples to make it even. Instead, here we can just assume that some processes will have a one fewer sample - so for e.g. 8 samples would be divided into 3, 3, 2 between three processes. This will also allow us to implement shuffling similar to the distributed sampler without any further changes.

Reply:
Re. shuffling, yes, for me we never need to shuffle in eval or prediction mode. If we want to shuffle we can just use the default DistributedSampler...

On the subsampling strategy, I was worried that if processes don't get the same number of samples they might not get the same number of _batches_, and then synchronization primitives in e.g. the forward passes will hang.

We might be able to solve this by better splitting the indices between processes, for instance for 7 samples in 3 buckets:

- Right now it would do 3 3 1 (bad)
- It would be better if it was 3 2 2 (better)

But even then there might be pathological cases where even a difference of one sample creates a difference in number of batches. 🤔

Reply:
@julien-c AFAICS currently the DistributedSampler as it is used in the trainer does not shuffle either. Well, that is to say that it only shuffles once, but that every following epoch has the same order. this is the case because the sampler deterministically shuffles based on the current epoch, which has to be set manually.

So the start of the epoch loop should look something like:

```python
for epoch in train_iterator:
    if isinstance(train_dataloader.sampler, DistributedSampler):
        train_dataloader.sampler.set_epoch(epoch)
```

Perhaps that is better suited for another PR?

Reply:
Oh yeah you're correct, will add to this PR